psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -888,24 +973,448 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+        int nQshell = shell_endpoints_for_atom[Qatom + 1] - shell_endpoints_for_atom[Qatom];
+
+        // First shell per atom
+        int Pstart = shell_endpoints_for_atom[Patom];
+        int Qstart = shell_endpoints_for_atom[Qatom];
+
+        // Number of basis functions per atom
+        int nPbasis = basis_endpoints_for_shell[Pstart + nPshell] - basis_endpoints_for_shell[Pstart];
+        int nQbasis = basis_endpoints_for_shell[Qstart + nQshell] - basis_endpoints_for_shell[Qstart];
+
+        int thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+
+        // => "Pre-ordering and Pre-selection to find significant elements in Puv" <= //
+        std::vector<std::vector<int>> significant_kets_P(nPshell);
+        std::vector<std::vector<int>> significant_kets_Q(nQshell);
+
+        // => Defined in Oschenfeld Eq. 3 <= //
+        // For shells P and R, If |Dpr| * sqrt(Pmax|Pmax) * sqrt(Rmax|Rmax) [screening value] >= linK_ints_cutoff_,
+        // Then shell R is added to the ket list of shell P (and sorted by the screening value)
+        for (size_t P = Pstart; P < Pstart + nPshell; P++) {
+            std::vector<std::pair<int, double>> PR_shell_values;
+            for (size_t R = 0; R < nshell; R++) {
+                double screen_val = shell_ceilings[P] * shell_ceilings[R] * ints[0]->shell_pair_max_density(P, R);
+                if (screen_val >= linK_ints_cutoff_) {
+                    PR_shell_values.emplace_back(R, screen_val);
+                }
+            }
+            std::sort(PR_shell_values.begin(), PR_shell_values.end(), screen_compare);
+
+            for (const auto& value : PR_shell_values) {
+                significant_kets_P[P - Pstart].push_back(value.first);
+            }
+        }
+
+        for (size_t Q = Qstart; Q < Qstart + nQshell; Q++) {
+            std::vector<std::pair<int, double>> QR_shell_values;
+            for (size_t R = 0; R < nshell; R++) {
+                double screen_val = shell_ceilings[Q] * shell_ceilings[R] * ints[0]->shell_pair_max_density(Q, R);
+                if (screen_val >= linK_ints_cutoff_) {
+                    QR_shell_values.emplace_back(R, screen_val);
+                }
+            }
+            std::sort(QR_shell_values.begin(), QR_shell_values.end(), screen_compare);
+
+            for (const auto& value : QR_shell_values) {
+                significant_kets_Q[Q - Qstart].push_back(value.first);
+            }
         }
+
+        // Keep track of contraction indices for stripeout (Towards end of this function)
+        std::vector<std::unordered_set<int>> P_stripeout_list(nPshell);
+        std::vector<std::unordered_set<int>> Q_stripeout_list(nQshell);
+
+        bool touched = false;
+        for (int P = Pstart; P < Pstart + nPshell; P++) {
+            for (int Q = Qstart; Q < Qstart + nQshell; Q++) {
+
+                int dP = P - Pstart;
+                int dQ = Q - Qstart;
```

Comment:
Since you need both P and Q as well as dP and dQ, it might be simpler to loop over dP and dQ instead. But this is a stylistic issue.